ref(rules): Migrate 'bad' rules to be disabled

ceorourke opened the pull request:
Migrate alert rules that have no action or are exact duplicates of another rule in the same project to be disabled. We now prevent these from being created, this will just clean up old ones. 

Closes #55502

ceorourke commented on src/sentry/migrations/0546_migrate_no_action_dupe_issue_alerts.py:
```diff
@@ -0,0 +1,81 @@
+
+
+def has_duplicate_rule(apps, rule_data, project, rule_id):
+    Rule = apps.get_model("sentry", "Rule")
```
Is this the right way to do it again? I need to compare the rule to other ones in the same project

ceorourke replied:
also this function is basically https://github.com/getsentry/sentry/blob/master/src/sentry/api/endpoints/project_rules.py#L38 but modified slightly for the migration to return a boolean since we don't need the duplicated rule

markstory replied:
Yes, this is the correct way to get a reference to model classes in migrations.

markstory commented on src/sentry/migrations/0546_migrate_no_action_dupe_issue_alerts.py:
```diff
@@ -0,0 +1,81 @@
+                    # one rule has the matcher and the other one doesn't
+                    keys += 1
+
+        if keys == matches:
```
Should this be inside the loop? The function could return on the first iteration of the loop.

ceorourke replied:
Isn't it inside the loop? 🤔

ceorourke replied:
Or are you saying it shouldn't be? All I care about is if it finds one match.

ceorourke replied:
Yep I just need to know if there is a duplicate at all

JoshFerge reviewed (commented):
is it possible to write a test for this? I know we previously added some tooling around data migration tests

ceorourke commented:
> is it possible to write a test for this? I know we previously added some tooling around data migration tests

I can - could you point me to where that is? I'm not finding much for `test_migration`, `/tests/migrations` etc.

JoshFerge commented:
> > is it possible to write a test for this? I know we previously added some tooling around data migration tests
> 
> I can - could you point me to where that is? I'm not finding much for `test_migration`, `/tests/migrations` etc.

Let me know if there's enough context here: https://develop.sentry.dev/database-migrations/#testing

ceorourke commented:
> > > is it possible to write a test for this? I know we previously added some tooling around data migration tests
> > 
> > 
> > I can - could you point me to where that is? I'm not finding much for `test_migration`, `/tests/migrations` etc.
> 
> Let me know if there's enough context here: https://develop.sentry.dev/database-migrations/#testing

I added a test based on that but running it locally (even with the env var set) always says: `Migration is no longer runnable. Retain until migration is removed.` 😞 This happens on almost every other test in the directory... might be because `0547` has kinda borked my local env.

ceorourke commented:
> > > > > is it possible to write a test for this? I know we previously added some tooling around data migration tests
> > > > 
> > > > 
> > > > I can - could you point me to where that is? I'm not finding much for `test_migration`, `/tests/migrations` etc.
> > > 
> > > 
> > > Let me know if there's enough context here: https://develop.sentry.dev/database-migrations/#testing
> > 
> > 
> > I added a test based on that but running it locally (even with the env var set) always says: `Migration is no longer runnable. Retain until migration is removed.` 😞 This happens on almost every other test in the directory... might be because `0547` has kinda borked my local env.
> > Update: Fixed 0547 locally and I get the same message.
> 
> ah annoying! i'll try running it on my local in a sec and see what happens. thanks for adding that!

Oh duh - they're all marked as skippable: `@pytest.mark.skip("Migration is no longer runnable. Retain until migration is removed.")` why is that though?

ceorourke commented on tests/sentry/migrations/test_0549_migrate_no_action_dupe_issue_alerts.py:
```diff
@@ -0,0 +1,49 @@
+        )
+
+    def test(self):
+        assert self.no_action_rule == ObjectStatus.ACTIVE
```
hang on I made some really silly mistakes here before I realized how to run it

JoshFerge commented:
> > > > > > is it possible to write a test for this? I know we previously added some tooling around data migration tests
> > > > > 
> > > > > 
> > > > > I can - could you point me to where that is? I'm not finding much for `test_migration`, `/tests/migrations` etc.
> > > > 
> > > > 
> > > > Let me know if there's enough context here: https://develop.sentry.dev/database-migrations/#testing
> > > 
> > > 
> > > I added a test based on that but running it locally (even with the env var set) always says: `Migration is no longer runnable. Retain until migration is removed.` 😞 This happens on almost every other test in the directory... might be because `0547` has kinda borked my local env.
> > > Update: Fixed 0547 locally and I get the same message.
> > 
> > 
> > ah annoying! i'll try running it on my local in a sec and see what happens. thanks for adding that!
> 
> Oh duh - they're all marked as skippable: `@pytest.mark.skip("Migration is no longer runnable. Retain until migration is removed.")` why is that though?

@markstory do you know on this one? do we generally make them skip after the PR is merged?

JoshFerge commented on src/sentry/migrations/0549_migrate_no_action_dupe_issue_alerts.py:
```diff
@@ -0,0 +1,81 @@
+    matchers = {key for key in list(rule_data.keys()) if key not in ("name", "user_id")}
+    extra_fields = ["actions", "environment"]
+    matchers.update(extra_fields)
+    existing_rules = Rule.objects.exclude(id=rule_id).filter(project=project, status=0)
```
```suggestion
    existing_rules = Rule.objects.exclude(id=rule_id).filter(project=project, status=ObjectStatus.ACTIVE)
```

ceorourke replied:
I intentionally did this in case when someone runs it in the future that enum isn't available - I think in the past like [here](https://github.com/getsentry/sentry/blob/master/src/sentry/migrations/0507_delete_pending_deletion_rules.py) I copied it over to avoid that - I'll update it here.

JoshFerge replied:
ahh got it -- could be worth just adding a comment. that makes sense!
